File: OneToOneCallSample/app/src/main/java/com/stringee/apptoappcallsample/common/SensorManagerUtils.java
package com.stringee.apptoappcallsample.common;

import android.annotation.SuppressLint;
import android.content.Context;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.os.PowerManager;

@SuppressLint("WakelockTimeout")
public class SensorManagerUtils implements SensorEventListener {
    private static volatile SensorManagerUtils instance;
    private SensorManager mSensorManager;
    private Sensor mProximity;
    private PowerManager powerManager;
    private PowerManager.WakeLock wakeLock;
    private final Context context;

    public static SensorManagerUtils getInstance(Context context) {
        if (instance == null) {
            synchronized (SensorManagerUtils.class) {
                if (instance == null) {
                    instance = new SensorManagerUtils(context);
                }
            }
        }
        return instance;
    }

    public SensorManagerUtils(Context context) {
        this.context = context.getApplicationContext();
    }

    public SensorManagerUtils initialize(String tag) {
        if (mSensorManager == null) {
            mSensorManager = (SensorManager) context.getSystemService(Context.SENSOR_SERVICE);
            if (mProximity == null) {
                mProximity = mSensorManager.getDefaultSensor(Sensor.TYPE_PROXIMITY);
            }
            mSensorManager.registerListener(this, mProximity, SensorManager.SENSOR_DELAY_NORMAL);
        }

        if (powerManager == null) {
            powerManager = ((PowerManager) context.getSystemService(Context.POWER_SERVICE));

            int screenLockValue = PowerManager.PROXIMITY_SCREEN_OFF_WAKE_LOCK;
            if (wakeLock == null) {
                wakeLock = powerManager.newWakeLock(screenLockValue, tag);
            }
        }
        return instance;
    }

    public void turnOn() {
        if (!wakeLock.isHeld()) {
            wakeLock.acquire();
        }
    }

    public void turnOff() {
        if (wakeLock.isHeld()) {
            wakeLock.release();
        }
    }


    public void releaseSensor() {
        if (mSensorManager != null) {
            mSensorManager.unregisterListener(this);
            mSensorManager = null;
        }

        if (wakeLock != null && wakeLock.isHeld()) {
            wakeLock.release();
            wakeLock = null;
        }

        if (instance != null) {
            instance = null;
        }
    }

    @Override
    public void onSensorChanged(SensorEvent sensorEvent) {
    }

    @Override
    public void onAccuracyChanged(Sensor sensor, int i) {

    }
}

File: Kotlin-OneToOneCallSample/app/src/main/java/com/stringee/kotlin_onetoonecallsample/common/SensorManagerUtils.kt
package com.stringee.kotlin_onetoonecallsample.common

import android.annotation.SuppressLint
import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.PowerManager

class SensorManagerUtils private constructor(private val applicationContext: Context) : SensorEventListener {
    private var mSensorManager: SensorManager? = null
    private var mProximity: Sensor? = null
    private var powerManager: PowerManager? = null
    private var wakeLock: PowerManager.WakeLock? = null

    fun initialize(tag: String?): SensorManagerUtils? {
        if (mSensorManager == null) {
            mSensorManager = applicationContext.getSystemService(Context.SENSOR_SERVICE) as SensorManager
            if (mProximity == null) {
                mProximity = mSensorManager!!.getDefaultSensor(Sensor.TYPE_PROXIMITY)
            }
            mSensorManager!!.registerListener(this, mProximity, SensorManager.SENSOR_DELAY_NORMAL)
        }
        if (powerManager == null) {
            powerManager = applicationContext.getSystemService(Context.POWER_SERVICE) as PowerManager
            val screenLockValue = PowerManager.PROXIMITY_SCREEN_OFF_WAKE_LOCK
            if (wakeLock == null) {
                wakeLock = powerManager!!.newWakeLock(screenLockValue, tag)
            }
        }
        return instance
    }

    @SuppressLint("WakelockTimeout")
    fun turnOn() {
        if (!wakeLock!!.isHeld) {
            wakeLock!!.acquire()
        }
    }

    fun turnOff() {
        if (wakeLock!!.isHeld) {
            wakeLock!!.release()
        }
    }

    fun releaseSensor() {
        if (mSensorManager != null) {
            mSensorManager!!.unregisterListener(this)
            mSensorManager = null
        }
        if (wakeLock != null && wakeLock!!.isHeld) {
            wakeLock!!.release()
            wakeLock = null
        }
        if (instance != null) {
            instance = null
        }
    }

    override fun onSensorChanged(sensorEvent: SensorEvent) {}
    override fun onAccuracyChanged(sensor: Sensor, i: Int) {}

    companion object {
        @Volatile
        private var instance: SensorManagerUtils? = null
        fun getInstance(context: Context): SensorManagerUtils {
            return instance ?: synchronized(this) {
                instance
                    ?: SensorManagerUtils(context.applicationContext).also {
                        instance = it
                    }
            }
        }
    }
}
